File: Java/app/src/main/java/com/github/shin1x1/jsonparser/parser/JsonValue.java
package com.github.shin1x1.jsonparser.parser;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

public sealed interface JsonValue {
    record Null() implements JsonValue {
    }

    record True() implements JsonValue {
    }

    record False() implements JsonValue {
    }

    record Number(BigDecimal value) implements JsonValue {
        public Number(java.lang.String value) {
            this(new BigDecimal(value));
        }
    }

    record String(java.lang.String value) implements JsonValue {
    }

    record Array(List<JsonValue> value) implements JsonValue {
    }

    record Object(Map<java.lang.String, JsonValue> value) implements JsonValue {
    }
}

File: Kotlin/src/main/kotlin/parser/JsonValue.kt
package parser

sealed interface JsonValue {
    object True : JsonValue {
        override fun toString(): kotlin.String = "true"
    }

    object False : JsonValue {
        override fun toString(): kotlin.String = "false"
    }

    object Null : JsonValue {
        override fun toString(): kotlin.String = "null"
    }

    data class Number(
        val value: Double,
    ) : JsonValue {
        override fun toString(): kotlin.String = "Number($value)"
    }

    data class String(
        val value: kotlin.String,
    ) : JsonValue {
        override fun toString(): kotlin.String = "String($value)"
    }

    data class Array(
        val value: List<JsonValue>,
    ) : JsonValue {
        override fun toString(): kotlin.String = "Array($value)"
    }

    data class Object(
        val value: Map<kotlin.String, JsonValue>,
    ) : JsonValue {
        override fun toString(): kotlin.String = "Object($value)"
    }
}
